drivers/hmis_external_apis/app/jobs/hmis_external_apis/ac_hmis/update_unit_availability_job.rb

```diff
@@ -8,36 +8,99 @@ module HmisExternalApis::AcHmis
   class UpdateUnitAvailabilityJob < ApplicationJob
     include HmisExternalApis::AcHmis::ReferralJobMixin
 
+    JOB_LOCK_NAME = 'hmis_external_update_unit_availability'.freeze
+
+    # @param data_source_id [Integer]
+    # @param force [Boolean]
+    def perform(data_source_id:, force: false)
+      with_locks do
+        projects = Hmis::Hud::Project.where(data_source_id: data_source_id)
+        projects.find_each do |project|
+          force_update(project) if force
+          project.external_unit_availability_syncs.dirty.preload(:unit_type, :user).each do |sync|
+            # sync.user made the most recent user changes, either to occupancy or unit inventory
+            sync_project_unit_type(
+              project: project,
+              unit_type: sync.unit_type,
+              user: sync.user || default_user,
+            )
+            # track sync version
+            sync.update!(synced_version: sync.local_version)
+          end
+        end
+      end
+    end
+
+    protected
+
+    def default_user
+      @default_user ||= Hmis::User.system_user
+    end
+
+    def force_update(project)
+      unit_types = project.units.preload(:unit_type).map(&:unit_type).compact.uniq
+      unit_types.each do |unit_type|
+        unit_type&.track_availability(project_id: project.id, user_id: default_user.id)
+      end
+    end
+
+    def with_locks
+      # lock job specific lock to prevent overlapping runs
+      Hmis::HmisBase.with_advisory_lock(JOB_LOCK_NAME, timeout_seconds: 0) do
+        # lock to synchronize with the project importer which may change available units
+        Hmis::HmisBase.with_advisory_lock(
+          HmisExternalApis::AcHmis::Importers::ProjectsImporter::JOB_LOCK_NAME,
+          timeout_seconds: 0,
+          shared: true,
+        ) do
+          yield
+          # FIXME ideally should report to dead man's snitch or equiv
+        end
+      end
+    end
+
     # determine current capacity at project and update external system
-    # @param project_id [Integer]
-    # @param unit_type_id [Integer]
-    # @param requested_by [String]
-    def perform(project_id:, unit_type_id:, requested_by:)
-      project = Hmis::Hud::Project.find(project_id)
+    # @param project [Hmis::Hud::Project]
+    # @param unit_type [Hmis::UnitType]
+    # @param user [Hmis::User]
+    def sync_project_unit_type(project:, unit_type:, user:)
+      my_name = self.class.name
       project_mper_id = mper.identify_source(project)
-      raise "mper id for not found Hmis::Hud::Project##{project_id}" unless project_mper_id
 
-      unit_type = Hmis::UnitType.find(unit_type_id)
       unit_type_mper_id = mper.identify_source(unit_type)
-      raise "mper id for not found for Hmis::UnitType##{unit_type_id}" unless unit_type_mper_id
+      unless unit_type_mper_id
+        Sentry.capture_message("#{my_name} mper id not found for Hmis::UnitType##{unit_type.id}")
```

Comment:
This may be sufficient (only sending to Sentry,) but if we want this to go to slack as well, it might be helpful to use `@notifer.ping` see lib/tasks/test.rake for some examples.

Reply:
@eanders sure, I can add that. Is it generally true that we should alert to Slack and Sentry for a recoverable error like this?

Reply:
It has been the practice.  @gigxz do you have a preference in this case?